client/modules/cm-tarnation/src/index.ts

```diff
@@ -0,0 +1,223 @@
+import {
+  Language,
+  LanguageSupport,
+  defineLanguageFacet,
+  LanguageDescription,
+  languageDataProp,
+  EditorParseContext
+} from "@codemirror/language"
+import { styleTags, Tag, tags } from "@codemirror/highlight"
+import type { Extension } from "@codemirror/state"
+
+import { removeUndefined } from "wj-util"
+import { Parser } from "./parser"
+import { Tokenizer } from "./tokenizer"
+import { Buffer, BufferCache } from "./buffer"
+import { Grammar } from "./grammar/grammar"
+import type * as DF from "./grammar/definition"
+
+import { Input, NodeProp, NodePropSource, NodeSet, NodeType, Tree } from "lezer-tree"
+import { isFunction } from "is-what"
+
+export * from "./grammar/helpers"
+
+// TODO: reuse ahead buffer rather than just always viewport skipping
+// TODO: fix string with variable array expansion (convert to array of codepoints)
+// TODO: add substitution to next, switchto
+// TODO: better document the grammar classes
+// TODO: better document the parser classes
+
+type AddNodeSpec = { name: string } & Omit<
+  Parameters<typeof NodeType["define"]>[0],
+  "id" | "name"
+>
+
+interface ParserConfiguration {
+  props?: NodePropSource[]
+}
+
+/** The options / interface required to create a Tarnation language. */
+export interface TarnationLanguageDefinition {
+  /** The name of the language.
+   *  This property is important for CodeMirror, so make sure it's reasonable. */
+  name: string
+  /** The grammar that will be used to tokenize the language.
+   *
+   *  This value can be provided as a function,
+   *  which will cause the grammar to be lazily evaluated. */
+  grammar: DF.Grammar | (() => DF.Grammar)
+  /** A list of `LanguageDescription` objects that will
+   *  be used when the parser nests in a language. */
+  nestLanguages?: LanguageDescription[]
+  /** Configuration options for the parser, such as node props. */
+  configure?: ParserConfiguration
+  /** A list of aliases for the name of the language. (e.g. 'go' -> ['golang']) */
+  alias?: string[]
+  /** A list of file extensions. (e.g. ['.ts']) */
+  extensions?: string[]
+  /** The 'languageData' field for the language.
+   *  CodeMirror plugins use this data to interact with the language. */
+  languageData?: Record<string, any>
+  /** Extra extensions to be loaded. */
+  supportExtensions?: Extension[]
+}
+
+/** Global handler for a Tarnation language.
+ *  The language constructed will not be processed until the `load` function is called.
+ *  @see TarnationLanguageDefinition */
+export class TarnationLanguage {
+  declare description: LanguageDescription
+  declare grammar?: Grammar
+  declare nodes?: NodeMap
+  declare support?: LanguageSupport
```

Comment:
it's weird how this isn't documented

Using declare here means that were just describing the type of a property of the class, rather than actually emitting an initializer.

```ts
class {
  foo: string
}
// emit
class {
  foo
}
```

```ts
class {
  declare foo: string
}
// emit
class {
}
```